### ISSUE TITLE
Allow to export const variables in gdscript

### ISSUE THREAD

**Issue by patrickelectric** [godotengine/godot-proposals#1165]
**Describe the project you are working on:**

Allow usage of the export keyword with const variables.
E.g: `export const fish = "tainha"`

**Describe the problem or limitation you are having in your project:**

As good practice immutable variables should be declared as const.
There is scenarios where the export of const variables are necessary,
like scripts that define const values that are exported for const mathematical
properties, dictionaries and etc.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**

It allows the usage of const variables with export keyword.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

Allows the usage of `export const fish = "tainha"`

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

The variable will be declared as mutable, and this is a bad practice for const variables.

In `globals.gd`
`export var fish = "tainha"`

In a different script, someone crazy can do things like:

`Globals.fish = "cardoso"`

**Is there a reason why this should be core and not an add-on in the asset library?:**

This need to be done in the gdscript parser.

**Comment by Calinou** [comment]
As of Godot 3.2, `const` in Godot is meant to be used for *compile-time* constants (like `constexpr` in C++).

That said, Godot 4.0 will support immutable variables using `const` at a local level (like `const` in C++). I don't know if the concept of compile-time constants will still be around.

**Comment by patrickelectric** [comment]
> If an `export` statement can change a `const`'s value, it's not a compile-time constant anymore (but just a run-time one).

That's something that is desired to avoid in this proposal and in the open PR.

**Comment by felaugmar** [comment]
At some point I wanted this feature, it's useful to have 'exportable consts' that can only be changed by the editor and in definition.

**Comment by Zylann** [comment]
If you have a `.gd` script and want `const` to be editable from the inspector, where would the new value be stored?
If you expect it to be next to other exported *variables*, it means it's not a `const`, it would become something else, tied to *instances of the script* (so not global), rather than the script itself, as `const` currently is.
If you expect it to be on the script resource itself, that means you can't do that with a `.gd` script, as there is no place to store the value in that "format".

**Comment by aaronfranke** [comment]
> exported for const mathematical properties

Mathematical constants should never be exported. There is no use case for changing the value of `Tau` for example.

**Comment by Diaspater** [comment]
@aaronfranke 
While mathematical constants like 'Tau' should never be overridden, a user may define a personal mathematical constant that they do not want to be overridden at runtime for what ever reason. For example, if I want movement speed to be 32 pixels per second, and define this speed as a 'const' to ensure that it isn't overridden at runtime. In addition because it is a constant I can use it in static functions and other scripts can reference it without needing an instance of the original script to do so. Now suppose that as the developer, I want to see what would happen if I changed that constant to 64 pixels per second or 12. Currently the only way to do so is to go into the script and manually change it to each of those values. Not impossible by any means, but an inconvenience. If it was allowed to be exported, I could simply use the inspector tool to change it to the desired value but it would still retain the value of not being able to be modified at runtime.

This isn't a perfect example as it likely violates some other design pattern best practices but I'm sure that there is a user case this proposal would be ideal for.

**Comment by LightningAA** [comment]
Duplicate of https://github.com/godotengine/godot-proposals/issues/820?

**Comment by dalexeev** [comment]
If you are talking about constants, then this is impossible, because the same script can be used by several instances - there is no main scene/resource for some script, all instances are equivalent. And constants are common to the class and all its instances.

![](https://user-images.githubusercontent.com/47700418/152590570-c2892428-b7a8-4983-9b43-d877c783f147.png)

If you need immutable variables, see LightningAA's comment above.

**Comment by Lexyth** [comment]
> If you are talking about constants, then this is impossible, because the same script can be used by several instances - there is no main scene/resource for some script, all instances are equivalent. And constants are common to the class and all its instances.
> 
> ![](https://user-images.githubusercontent.com/47700418/152590570-c2892428-b7a8-4983-9b43-d877c783f147.png)
> 
> If you need immutable variables, see LightningAA's comment above.

That is true for exported _variables_, since they behave like _variables_ in a script.
It doesn't have to be true for exported _constants_, since they could be made to behave like consts in a script.
I.e.: Changing a const in one instance changes it globally and all instances just use that global value.
Or in reference to your example image, b would just be 3 for all those tscn/tres files because the value is read/written to a shared place.
I'm not sure exactly how exporting variables works internally, but I'd wager whatever mechanism keeps track of values for each instance also has a reference to the class and could be adapted to include a reference to a mechanism that keeps track for constants of that class, too.

**Comment by dalexeev** [comment]
There is no centralized storage for constant values and scripts do not support metadata. Even if we somehow implemented `@export const`, it would be a completely different implementation, despite the same annotation. The mechanism used to export properties is not suitable for constants.

However, you can use a preloaded resource to initialize a constant, the following code works:

```gdscript
const MY_VALUE = preload("resource.tres").my_value
```

But I don't quite understand the point of being able to edit constants ​​via the inspector rather than via code. If you expect that changing a constant will instantly apply everywhere, then it doesn't. Constants in GDScript are compile time evaluated, so this would require recompiling all scripts that use the constant (and probably all dependent scripts, transitively).

**Comment by Lexyth** [comment]
Not everyone working with Godot is a programmer. Maybe a level designer wants to see how the map looks with 200 instead of 50 randomly placed objects and that values is defined as a constant.

As for the compile time aspect, I don't see a difference to changing it in code, other than that changing the constant in the editor will inherently not be reflected at runtime like an exported var would, which is desirable for a const. If changes are meant to be reflected at runtime, one should use a var to begin with.

All exported const fields could just be disabled in the editor during runtime. 
And on changes, the script is recompiled exactly the same as if the code itself had been changed.
A central store could just consist of an internal resource instance that is equal for all resources of that class, a "main resource" (although a dedicated system would be better...).
Exported consts values will be replaced with the values from the central store instead of the value that they are assigned in code, if it exists, else not (this would be the same as with exported vars, except for the redirecting to the central store).
The editor then just reads and writes the value to and from this resource as a reaction to reading and writing to the current resource (whose values for that const would just be ignored anyway), similar to how input validation works.

I'm sure this isn't perfect, nor very elegant, but there are already systems in place for the individual steps, so it should be implementable this way without having to create a dedicated system, unless I'm misunderstanding something.

**Comment by trevorbaughn** [comment]
> scripts that define const values that are exported for const mathematical properties

Exactly my use-case building a snowboarding game.  There are way too many things to keep track of that should be constants, 20 of them so far, and I'd love to be able to export for designers... but coming across this issue I see this isn't a thing yet.

**Comment by Zireael07** [comment]
@trevorbaughn In my space game I had a similar use case. I ended up putting them as constants in an autoload

**Comment by WinterAtne** [comment]
> If you are talking about constants, then this is impossible, because the same script can be used by several instances - there is no main scene/resource for some script, all instances are equivalent. And constants are common to the class and all its instances.
> 
> ![](https://user-images.githubusercontent.com/47700418/152590570-c2892428-b7a8-4983-9b43-d877c783f147.png)
> 
> If you need immutable variables, see LightningAA's comment above.

I understand completely that the current consts are compile time and the same across all scripts using the constant, however, exported constants, or something similar do not necessarily have to work the same way. The desire is simply to have values that are mutable when designing (by assigning the value in the inspector), but immutable at runtime. Those asking for this change (including myself), are not looking for the optimization brought by compile time constants, just the security of unchanging values during program execution.

I do recognize that this would likely cause a lot of problems in the parser, however if that proves to be a significant barrier, we could introduce a new keyword like ```val``` as mentioned in https://github.com/godotengine/godot-proposals/issues/820

**Comment by trevorbaughn** [comment]
I agree with the above.

Not understanding the engine architecture myself though - would it not be possible to have them function differently in the editor before being compiled for runtime, effectively keeping runtime optimization?  I feel some sort of way of notifying the user that it'll be changed in X different places may be necessary still... which itself could be a massive pain.  I also understand this sort of difference with runtime is not typically good... but as described above - the main desire is for inspector-editable constants for designers that do not wish to dig through code in the file system to edit before runtime, while still getting the security of unchanging values during execution.  So long as the differences are documented or otherwise made clear, I don't think users would mind?

**Comment by ssokolow** [comment]
\*nod* The workflow I imagine for `export const` is that they appear in the editor but the editor widget's Editable property is forcibly toggled to Off while the game is running so that, as far as a user is concerned, it's just a fancier UI for editing a const in the source before running it.

...the kind of thing that, if someone were to pull a [Fake Go generics before it had them using Canadian Aboriginal Syllabics and a preprocessor](https://github.com/vasilevp/aboriginal), would be accomplished something like:

1. The user uses the external editor option in Godot
2. The actual source is some kind of `.gd.in` file containing substitution tokens
3. The user's editor has a plugin which automatically redirects attempts to open the `.gd` file to opening the `.gd.in` file instead, and automatically invokes a macro preprocessor to do  TypeScript-esque compilation on saving the `.gd.in`.
4. The preprocessor pulls the values for the substitution tokens from something like a TOML file and there's a custom editor plugin in the Godot project which exposes a GUI for editing the values in the TOML file.
